src/embed_tests/TestPyScope.cs

```diff
@@ -337,9 +338,12 @@ public void TestThread()
                     //add function to the scope 
                     //can be call many times, more efficient than ast 
                     ps.Exec(
+                        "import clr\n" +
+                        "from System.Threading import Thread\n" +
                         "def update():\n" +
                         "    global res, th_cnt\n" +
                         "    res += bb + 1\n" +
+                        "    Thread.MemoryBarrier()\n" +
```

Comment:
@lostmsu  Not a good idea for using some clr code here, and the counter didn't matched were because `res += bb + 1` it's not a atomic operate, they had been separated to several op codes. So even make some barrier here it may still be failed(if pushing more theads), but decrease the probability.
thread 1: executing `bb + 1` before assign to `res`
thread 2: executing op `STORE_FAST` of `res += `
context changed between these two threads, thus res may only increase once.
this should be enough: https://github.com/pythonnet/pythonnet/pull/958/commits/da97502006791bb0597446766ad00a6f9d291895

Reply:
Regarding CLR: Python does not have a defined memory model, and I could not find any primitive analogous `Thread.MemoryBarrier` in Python. `threading.Lock` that you suggested does not have any mentions of memory barriers, so I could not rely on it.

Regarding atomicity. I think your understanding of the issue is incorrect. All calls to `update` functions from .NET threads are performed while having GIL (they are wrapped into `using Py.GIL()`), so only 1 thread executes the entirety of `update` at once and from this standpoint of view it should be considered atomic.

The issue is that acquiring GIL (at least judging by Python 2.7 code) does not introduce any kind of memory barrier, so after .NET thread 1 finished executing `update`, and .NET thread 2 entered it, thread 2 might not see the changes thread 1 made, even though thread 1 already completed and exited the `update` function.

Reply:
That sounds right, but even so, it still needs the lock, change code like this, you shall see.
![图片](https://user-images.githubusercontent.com/3406561/75466620-6e2b9200-59c5-11ea-89dd-7bdce6f2f605.png)

Reply:
Looks like you are right, memory barrier is insufficient. I though GIL ensures only 1 Python thread can execute while it is held, but actually it just ensures, the the current thread can execute Python code, and more than 1 can do so simultaneously. Now I wonder if we should simply lock in this test on .NET side. But I guess it does not matter, if you do it in Python or in C#. Might still need a memory barrier if Python's lock does not insert one.

Reply:
Opened https://github.com/pythonnet/pythonnet/issues/1067 to track

Reply:
Have you considered a race condition? The `PyScopeTest` is not thread safe, in general. It is set up to have a single private `PyScope` at the class level that is created/disposed during `SetUp`/`Dispose`, and referenced in each test. However, tests within the fixture can, and do, run in parallel. 
So, `ps` can be modified mid-test for any test in this fixture. `TestThread` is one of the longer tests in the fixture, so one might expect it to be more vulnerable to the data races expected of a shared state variable being replaced.
Have you considered setting either setting `NonParallelizable` on all of the tests in the fixture, or making the fixture itself threadsafe by making each test create its own, locally scoped, `PyScope`?